### PR TITLE
Fix event schemas for /sync

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check

----

This commit clarifies the required keys for each type of event that appears in sync, fixes the core event schema not declaring 'content' as required, and includes a mention that events may not have a room_id when appearing in /sync.

Fixes https://github.com/matrix-org/matrix-doc/issues/595
Fixes https://github.com/matrix-org/matrix-doc/issues/909

### CODE REVIEW

File: event-schemas/schema/core-event-schema/sync_room_event.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+allOf:
+- $ref: event.yaml
+description: In addition to the Event fields, Room Events have the following additional
+  fields.
+properties:
+  event_id:
+    description: The globally unique event identifier.
+    type: string
+  sender:
+    description: Contains the fully-qualified ID of the user who *sent*
```
Comment (ara4n) [inline event-schemas/schema/core-event-schema/sync_room_event.yaml]: why the asterisks? i found myself distracted and wondering what the sublety was.

File: event-schemas/schema/core-event-schema/sync_state_event.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+properties:
+  prev_content:
+    description: Optional. The previous ``content`` for this event. If there is no
+      previous content, this key will be missing.
+    title: EventContent
+    type: object
+  state_key:
+    description: A unique key which defines the overwriting semantics for this piece
+      of room state. This value is often a zero-length string. The presence of this
+      key makes this event a State Event. The key MUST NOT start with '_'.
```
Comment (ara4n) [inline event-schemas/schema/core-event-schema/sync_state_event.yaml]: hm, why is _ special?

File: event-schemas/schema/core-event-schema/event.yaml [hunk 3]
```diff
@@ -10,5 +10,6 @@ properties:
     type: string
 required:
   - type
+  - content
```
Comment (ara4n) [inline event-schemas/schema/core-event-schema/event.yaml]: is `content` really required? i think we consciously made it optional on the basis that you could have an event type that just had no content beyond its type (although I agree it's a bit stupid, and I don't have a problem with mandating it, so long as we've vaguely understood why it was optional before)

File: event-schemas/schema/core-event-schema/sync_room_event.yaml (hunk 1 above)
Comment (turt2live) [reply event-schemas/schema/core-event-schema/sync_room_event.yaml]: This is copy/pasted from the room_event.yaml - happy to remove them though.

File: event-schemas/schema/core-event-schema/sync_state_event.yaml (hunk 2 above)
Comment (turt2live) [reply event-schemas/schema/core-event-schema/sync_state_event.yaml]: also copy-pasted from it's parent file. It's not special as far as I know - will remove it.

File: event-schemas/schema/core-event-schema/event.yaml (hunk 3 above)
Comment (turt2live) [reply event-schemas/schema/core-event-schema/event.yaml]: I can't think of an event that has no content, and this doesn't require the object to actually contain anything. I honestly thought it was a mistake that it wasn't included and expect that client authors (for instance) would assume that the field exists, at their own detriment.

Comment (richvdh) [reply event-schemas/schema/core-event-schema/event.yaml]: We do enforce its presence when submitting messages over the C-S API in synapse. +1 to requiring its presence here.

**Comment by turt2live** [comment]
@richvdh please take a quick look at the recent changes - had to make some corrections to ensure the project builds.

**Pull request merged**